oldw <- getOption("warn")
options(warn = -1)

suppressPackageStartupMessages(library(optparse))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(dplyr))

option_list <- list( 
  make_option(c("-Q", "--eqtl"), action="store",
              help="Tissue-specific significant gene-eQTL pairs"),
  make_option(c("-S", "--snv"), action="store",
              help="Germline SNV genotypes for all samples"),
  make_option(c("-R", "--results"), action="store",
              help="HYENA results file '_posEstimate.txt' or '_adj.txt' or'_adj_sig.txt'"),
  make_option(c("--emp"), action="store_true", default=FALSE,
              help="P values are empirical p-values ('p.emp.fdr')"),
  make_option(c("-f", "--fdrcutoff"), action="store", default= 0.1, 
              help="FDR cutoff for genes [default %default]"),
  make_option(c("-c", "--sigcutoff"), action="store", default= 0.05, 
              help="P-value cutoff for eQTLs [default %default]"),
  make_option(c("-g", "--genome"), action="store", 
              help="Reference genome version, 'hg19' or 'hg38'"),
  make_option(c("-x", "--prefix"), action="store", default="DAT",
              help="Prefix for output files [default %default]"),
  make_option(c("-V", "--svstat"), action="store",
              help="*sv_mapped_filtered_numSV.txt generated by mapsv.R"),
  make_option(c("-i", "--id"), action="store",
              help="Sample id file"),
  make_option(c("-a", "--annot"), action="store", 
              help="Gene annotation with TSS window"),
  make_option(c("-w", "--write"), action="store", default="./intermediate/",
              help="Folder for intermediate files [default %default]"),
  make_option(c("-d", "--dir"), action="store", default= "./results/", 
              help="Output directory for the results. If no path is entered, a new directory called %default will be created automatically"),
  make_option(c("-v","--verbose"), action="store_true", default=FALSE,
              help="Verbose mode for troubleshooting [default %default]")
)

parser <- OptionParser(usage="%prog [options]", option_list=option_list, description="")
args <- parse_args(parser, positional_arguments=0)
opt <- args$options

# Load data
eqtl <- fread(opt$eqtl, stringsAsFactors = FALSE)
snv <- fread(opt$snv, stringsAsFactors = FALSE)
results <- fread(opt$results, stringsAsFactors = FALSE)

# Determine significant genes
if(opt$emp){
  sig <- results[results$p.emp.fdr <= opt$fdrcutoff,]
  } else {
  sig <- results[results$fdr.onesided <= opt$fdrcutoff,]
  }


# Extract eqtls for significant genes
eqtl <- eqtl[(vapply(strsplit(eqtl$gene_id,"\\."), `[`, 1, FUN.VALUE=character(1))) %in% (vapply(strsplit(sig$gene_id,"\\."), `[`, 1, FUN.VALUE=character(1))),]

# Merge data based on hg19 (you might want to provide option for hg38 as well here)
if (opt$genome == "hg19") {
  genotypes <- merge(eqtl, snv, all.x=TRUE, by.x="variant_id_b37", by.y="variant_id") 
} else if(opt$genome == "hg38") {
  genotypes <- merge(eqtl, snv, all.x=TRUE, by.x="variant_id_b38", by.y="variant_id")
} else {
  print("This version of reference genome is currently not supported by eqtl.R")
}


if (file.exists(opt$write)) {
 write.table(genotypes, file = paste0(opt$write, opt$prefix, ".sig_genotypes.txt"), quote = FALSE, sep = "\t")
} else {
 dir.create(opt$write)
 write.table(genotypes, file = paste0(opt$write, opt$prefix, ".sig_genotypes.txt"), quote = FALSE, sep = "\t")
}


#### TEST eQTL vs SV genotype
svstat <- fread(opt$svstat, header = F, stringsAsFactors = F)
svstat <- t(svstat) # transpose
colnames(svstat) <- svstat[1,] # make the ENSG names the column names
svstat <- svstat[-1,] # remove first row
svstat <- as.data.frame(svstat, stringsAsFactors = T)
svstat <- slice(svstat, 1:(n() - 2))  # remove last two rows

samples <- fread(opt$id, stringsAsFactors = F)
ref <- fread(opt$annot, stringsAsFactors = F)
#class(svstat) <- "data.frame"
class(samples) <- "data.frame"
class(ref) <- "data.frame"

genotypes <- na.omit(genotypes)
genotypes$gene_id <- vapply(strsplit(genotypes$gene_id,"\\."), `[`, 1, FUN.VALUE=character(1))
if(opt$verbose){print(head(genotypes$gene_id))}

colnames(svstat) <- vapply(strsplit(colnames(svstat),"\\."), `[`, 1, FUN.VALUE=character(1))
colnames(svstat)[1] <- "aliquot_id_wgs"
if(opt$verbose){print(head(colnames(svstat)))}

ref$gene_id <- vapply(strsplit(ref$gene_id,"\\."), `[`, 1, FUN.VALUE=character(1))

key <- c("0|0"="AA", "0/0"="AA", "0|1"="AB", "0/1"="AB", "1|0"="AB", "1/0"="AB", "1|1"="BB", "1/1"="BB" )  

gene_id <- c()
variant_id <- c()
pvalue <- c()

for (i in 1:nrow(genotypes)){
  gene <- genotypes[i,]$gene_id
  var <- genotypes[i,]$variant_id_b37
  
  if(!(gene %in% colnames(svstat))){next} 

  temp <- t(genotypes[i,23:ncol(genotypes)])
  temp <- cbind(temp,rownames(temp))
  colnames(temp) <- c("genotype","aliquot_id_wgs_norm")
  temp <- merge(samples, temp, by = "aliquot_id_wgs_norm", all.y = T)
  temp$genotype <- key[temp$genotype]
  
  sv_status <- svstat[,c("aliquot_id_wgs", gene)]
  temp <- merge(temp, sv_status, by="aliquot_id_wgs", all.x = T)
  
  if(opt$verbose){print(paste0("Genotype at ",var, " for the gene ", gene, ":"))}
  if(nlevels(as.factor(temp$genotype))==1){next}
  if(nlevels(as.factor(temp[,gene]))==1){next}
  if(opt$verbose){print(table(temp[,gene], temp[,"genotype"]))}
  p <- chisq.test(temp[,gene], temp[,"genotype"], correct = F)$p.value
  pvalue <- c(pvalue, p)
  gene_id <- c(gene_id, gene)
  variant_id <- c(variant_id, var)
}

chisq_results <- cbind(variant_id, gene_id)
chisq_results <- cbind(chisq_results, pvalue)
chisq_results <- merge(chisq_results, ref, by="gene_id", all.x=T)

write.table(chisq_results, file = paste0(opt$write, opt$prefix, ".chisq_results.txt"), quote = FALSE, sep = "\t", row.names = FALSE)


# filter out genes with eQTL associations
chisq_results <- chisq_results[chisq_results$pvalue <= opt$sigcutoff,]
genes <- unique(chisq_results$gene_id)

sig <- sig[!((vapply(strsplit(sig$gene_id,"\\."), `[`, 1, FUN.VALUE=character(1))) %in% genes),]


cond <- strsplit(opt$results, "/")
cond <- cond[[1]][length(cond[[1]])]
cond <- gsub(".txt","", cond)

if (!(file.exists(opt$dir))) {
  dir.create(opt$dir, recursive = TRUE)
}
write.table(sig, file = paste0(opt$dir, cond, "_rmeqtl.txt"), quote = FALSE, sep = "\t", row.names = FALSE)



options(warn = oldw)
