#!/bin/env Rscript
oldw <- getOption("warn")
options(warn = -1)


suppressPackageStartupMessages(library(optparse))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(stats))
suppressPackageStartupMessages(library(MatrixGenerics))
suppressPackageStartupMessages(library(ggplot2))


option_list <- list(
  make_option(c("-e", "--exp"), action="store",
              help="Expression Rdata file generated by add_noise.R or pmt_exp.R"),
  make_option(c("-s", "--sv"), action="store", 
              help="*sv_mapped_filtered_numSV.txt file generated by mapsv.R"),
  make_option(c("-i", "--id"), action="store", 
              help="Sample id file"),
  make_option(c("-a", "--annot"), action="store", 
              help="Gene annotation with TSS window"),
  make_option(c("-p", "--purity"), action="store", 
              help="Sample purity file"),
  make_option(c("-c", "--cna"), action="store",
              help="Gene level copy number file"),
  make_option(c("-m", "--clinical"), action="store", 
              help="Clinical file"),
  make_option(c("-n", "--npc"), action="store", type="integer", default= 0, 
              help="Number of first n PCs to be included in the model [default %default] if --PC argument is entered"),
  make_option(c("-d", "--dir"), action="store", default= "./results/", 
              help="Output folder"),
  make_option(c("-w", "--write"), action="store", default="./intermediate/",
              help="Folder for intermediate files [default %default]"),
  make_option(c("-x", "--prefix"), action="store", default="DAT",
              help="Prefix for output files [default %default]"),
  make_option(c("-v","--verbose"), action="store_true", default=FALSE,
              help="Verbose mode for troubleshooting [default %default]"),
  make_option(c("-t","--pmt"), action="store_true", default=FALSE,
              help="Run in permutation mode [default %default]"),
  make_option(c("--pur"), action="store_true", default=FALSE,
              help="Include purity in regression"),
  make_option(c("--cn"), action="store_true", default=FALSE, 
              help="Include copy number in regression"),
  make_option(c("--age"), action="store_true", default=FALSE,
              help="Include age in regression"),
  make_option(c("--sex"), action="store_true", default=FALSE,
              help="Include sex in regression"),
  make_option(c("--type"), action="store_true", default=FALSE,
              help="Include cancer type in regression"),
  make_option(c("--PC"), action="store_true", default=FALSE,
              help="Include principal components in regression"),
  make_option(c("-f", "--fcutoff"), action="store", type="integer", default= 5, 
              help="SV Frequncy (%) cutoff [default %default]"),
  make_option(c("-C", "--cncutoff"), action="store", type="integer", default= 10, 
            help="Maximum copy number cutoff [default %default]"),
  make_option(c("-g", "--genes"), action="store", default=FALSE,
              help="A list of genes of interest to be tested")
)

parser <- OptionParser(usage="%prog [options]", option_list=option_list, description="")
args <- parse_args(parser, positional_arguments=0)
opt <- args$options


# Load the input expression, SV, purity, CNA, and patient data ####
load(opt$exp)

if(opt$verbose){
  print("expression data loaded")
  print(fpkm.qn.t[1:5,1:5])
}

dat.svmapped <- fread(opt$sv, stringsAsFactors = FALSE)

if(opt$verbose){
  print("gene SV status loaded")
  print(dat.svmapped[1:5,1:5])
}

samples <- fread(opt$id, stringsAsFactors = FALSE)
if(opt$verbose){
  print("sample information loaded")
  print(head(samples))
}

transcripts <- fread(opt$annot, stringsAsFactors = FALSE)

if(opt$verbose){
  print("reference annotation loaded")
  print(head(transcripts))
}

if(opt$pur) {
  purity <- read.table(opt$purity, stringsAsFactors = FALSE, header = TRUE)
  if(opt$verbose){
    print("purity data loaded")
    print(head(purity))
  }
}

if(opt$cn) {
  cna <- fread(opt$cna, stringsAsFactors = FALSE)
  if(opt$verbose){
    print("copy number data loaded")
    print(cna[1:5,1:5])
  }
}

if(opt$age | opt$sex | opt$type) {
  clindat <- fread(opt$clinical, stringsAsFactors = FALSE)
  if(opt$verbose){
    print("clinical data loaded")
    print(head(clindat))
  }
}

if(opt$genes != FALSE) {
  goi <- read.table(opt$genes, quote="\"", comment.char="")
  goi <- goi$V1
  
  if(opt$verbose){
    print("gene list loaded")
    print(head(goi))
  }
}

# Prepare the SV data ####
if(opt$verbose){print("SV status matrix is being processed...")}
# prep and transpose the data
dat.svmapped <- t(dat.svmapped) # transpose

colnames(dat.svmapped) <- dat.svmapped[1,] # make the ENSG names the column names
dat.svmapped <- dat.svmapped[-1,] # remove first row

if(opt$verbose){
  print("SV data dimensions before aliquot id filter:")
  print(dim(dat.svmapped))
}

# match the order of rows in dat.svmapped to the order of samples in dat.expression according to the samples table
aliqid_wgs <- samples$aliquot_id_wgs
if(opt$verbose){print(head(aliqid_wgs))}
aliqid_rnaseq <- samples$aliquot_id_rnaseq
if(opt$verbose){print(head(aliqid_rnaseq))}

fpkm.qn.t <- fpkm.qn.t[match(aliqid_rnaseq, rownames(fpkm.qn.t)),]
fpkm.qn.t.ns <- fpkm.qn.t.ns[match(aliqid_rnaseq, rownames(fpkm.qn.t.ns)),]
dat.svmapped <- dat.svmapped[match(aliqid_wgs, rownames(dat.svmapped)),]
dat.svmapped <- as.data.frame(dat.svmapped, stringsAsFactors = FALSE)
rownames(dat.svmapped) <- gsub("X", "", rownames(dat.svmapped))
rownames(dat.svmapped) <- gsub("\\.", "-", rownames(dat.svmapped))

if(opt$verbose){
  print("SV data dimensions after aliquot id filter:")
  print(dim(dat.svmapped))
}

# only keep the genes that have expression values
genes <- colnames(dat.svmapped)
keep <- genes[genes %in% colnames(fpkm.qn.t)]
dat.svmapped <- dat.svmapped[,keep]

if(opt$verbose){
  print("SV data dimensions after gene expression filter:")
  print(dim(dat.svmapped))
}

if(opt$verbose){
  print("SV data dimensions after SV status filter:")
  print(dim(dat.svmapped))
  print("SV status matrix processed")
}


#if (file.exists(opt$write)) {
# write.table(dat.svmapped, file = paste0(opt$write, opt$prefix, ".dat_svmapped.txt"), quote = FALSE, sep = "\t")
#} else {
# dir.create(opt$write, recursive=TRUE)
# write.table(dat.svmapped, file = paste0(opt$write, opt$prefix, ".dat_svmapped.txt"), quote = FALSE, sep = "\t")
#}

# Prepare the purity data ####
if(opt$pur) {
  purity <- purity[purity$samplename %in% aliqid_wgs,]
  purity <- purity[match(aliqid_wgs, purity$samplename),]
  if(opt$verbose){
    print("purity data processed:")
    print(head(purity))}
}

# Prepare the CNA data ####
if(opt$cn) {
	cna <- as.matrix(cna)
	rownames(cna) <- cna[,1]
	cna <- cna[,-1]
	class(cna) <- "numeric"
	cna <- cna[,colnames(cna) %in% aliqid_wgs]
  cna <- cna[,match(aliqid_wgs, colnames(cna))]
	if(opt$verbose){
    print("cna processed:")
    print(cna[1:5,1:5])}
}
#write.table(cna, file = paste0(opt$write, opt$prefix, ".cna_processed.txt"), quote = FALSE, sep = "\t")

# Prepare the Clinical data ####
if(opt$age | opt$sex | opt$type) {
  clindat <- merge(clindat, samples, by = "submitter_donor_id")
  clindat <- clindat[match(aliqid_wgs, clindat$aliquot_id_wgs),]
  if(opt$verbose){
    print("clindat processed:")
    print(head(clindat))}
}

# more on PCA
pca.ind <- as.data.frame(pca.prcomp$x)
pca.ind <- pca.ind[match(aliqid_rnaseq, rownames(pca.ind)),]

if(!(opt$pmt)){
  if (file.exists(opt$write)) {
		write.table(pca.ind, file = paste0(opt$write, opt$prefix, ".pca_ind.txt"), quote = FALSE, sep = "\t")
		} else {
 		dir.create(opt$write, recursive=TRUE)
		write.table(pca.ind, file = paste0(opt$write, opt$prefix, ".pca_ind.txt"), quote = FALSE, sep = "\t")
	}
}


# Test PCs against the SV variable and remove PCs that correlate with SV status based on two-sided ranksum test ####
if (!(file.exists(paste0(opt$write, "matrix")))) {dir.create(paste0(opt$write, "matrix"), recursive=TRUE)}

Gene <- c()
Freq <- c()
Ratio <- c()
Estimate <- c()
StdError <- c()
t <- c()
pvalue <- c()

if(opt$genes != FALSE) {
  genelist <- goi
} else {
  genelist <- colnames(dat.svmapped)
}

print("Starting normal-score regression...")

for (i in 1:length(genelist)) {
  
  if (i %% 100 == 0){print(paste0("processing gene # ", i))}

  gene_id <- genelist[i]
  if(opt$verbose){print(paste0("Testing gene ", gene_id, ": i = ", i))}
  
  sv_status <- dat.svmapped[,gene_id]
  #print(paste0("Gene ", gene_id, " levels: ",levels(as.factor(sv_status))))
  #print(sv_status)
  
  check.pca <- function(x) {
    test <- cbind(sv_status,x)
    test <- as.data.frame(test)
    test$sv_status <- as.factor(test$sv_status)
    #print(test)
    result <- wilcox.test(as.numeric(as.character(test$x)) ~ test$sv_status, data = test)
    result$p.value
  }
  
  test.result <- apply(pca.ind, 2, check.pca)
  
  keep <- test.result[!(test.result <= 0.1)] # vector with PCs that DO NOT significantly correlate with the SV variable
  keep <- names(keep)
  pca.ind.temp <- pca.ind[,keep]
  
  
  # Perform generalized linear model while correcting for principal components ####
  test.mat <- cbind(fpkm.qn.t.ns[,gene_id], sv_status)
  test.mat <- as.data.frame(test.mat)
  colnames(test.mat)[1] <- "expression"
  test.mat$expression <- as.numeric(as.character(test.mat$expression))
  #print("Initial testing matrix:")
  #print(head(test.mat))
  
  if(opt$pur) {
    pur <- purity$purity
    if(length(levels(as.factor(pur))) >= 2) {
      test.mat$pur <- as.numeric(as.character(pur))
    }
  }
  #print("Purity added to the testing matrix:")
  #print(head(test.mat))
  
  if(opt$cn) {
    if(gene_id %in% rownames(cna)) {
      cn <- as.numeric(as.character(cna[gene_id,]))
      if(!is.na(sum(cn, na.rm = TRUE))) { # do not include cn in the model if all "NA"s
        #print(paste0("Number of levels more than 2? ", (length(levels(as.factor(cn))) >= 2)))
        if (length(levels(as.factor(cn))) >= 2) {
          test.mat$cn <- cn
        }
      }
    }
  }
  #print("Copy number added to the testing matrix")
  #print(head(test.mat))
  
  if(opt$age) {
    age <- as.numeric(as.character(clindat$donor_age_at_diagnosis))
    if(length(levels(as.factor(age))) >= 2) {
      test.mat$age <- as.numeric(as.character(age))
    }
  }
  #print("Age added to the testing matrix")
  #print(head(test.mat))
  
  if(opt$sex) {
    sex <- clindat$donor_sex
    if(length(levels(as.factor(sex))) >= 2) {
      test.mat$sex<- sex
    }
  }
  #print("Sex added to the testing matrix")
  #print(head(test.mat))
  
  if(opt$type) {
    cancer_type <- clindat$cancer_type
    if(length(levels(as.factor(cancer_type))) >= 2) {
      test.mat$cancer_type<- cancer_type
    }
  }
  #print("Cancer type added to the testing matrix")
  #print(head(test.mat))
  
  if(opt$PC) {
    n <- opt$npc # number of first n PCs to be included in the linear model
    if (n == 1)
    {
   		test.mat <- cbind(test.mat, pca.ind.temp[,1])
   		colnames(test.mat)[length(colnames(test.mat))] <- colnames(pca.ind.temp)[1]
   	}
    if (n > 1)
    {
    	test.mat <- cbind(test.mat, pca.ind.temp[,1:n])
    }
  }
  #print("PCs added to the testing matrix")
  #print(test.mat)
  
  if(opt$verbose){
    print("Dimensions of the testing matrix before NA removal:")
    print(dim(test.mat))
    }
  
  test.mat2 <- na.omit(test.mat) # remove all rows with NA
  test.mat2 <- test.mat2[test.mat2$expression != "Inf",] # remove patients with Inf
  if(opt$verbose){
    print("Dimensions of the testing matrix after NA removal:")
    print(dim(test.mat2))
  }
  
  if(opt$verbose){
    print("Dimensions of the testing matrix after recurrence filter:")
    print(dim(test.mat2))
    }

  # Apply CN cutoff
  if(opt$cn) {
    if("cn" %in% colnames(test.mat2)){
      if(nrow(test.mat2[test.mat2$cn <= opt$cncutoff,]) >= 10) {
        test.mat2 <- test.mat2[test.mat2$cn <= opt$cncutoff,]
      }
    }
  }
  if(opt$verbose){
    print("Dimensions of the testing matrix after high CN filter:")
    print(dim(test.mat2))
    }
	
	# check levels of sex again
  if(opt$sex)
  {
    if(length(levels(as.factor(table(test.mat2$sex)))) < 2)
  	{
  		if("sex" %in% colnames(test.mat2))
  		{
  			test.mat2 <- subset(test.mat2, select = -c(sex))
  		}
 		}
  }
  
  #print("Processed testing matrix:")
  #print(test.mat2)

  # Calculate sv ratio and frequency
  if(length(which(test.mat2$sv_status %in% "no_sv")) == 0) {
    no_sv = 0
  } else if(length(which(test.mat2$sv_status %in% "sv")) == 0) {
    sv = 0
  } else {
    sv <- length(which(test.mat2$sv_status %in% "sv"))
    no_sv <- length(which(test.mat2$sv_status %in% "no_sv"))
  }
  
  # at least 3 samples in both sv and no_sv groups
  if(length(levels(as.factor(table(test.mat2$sv_status)))) < 2)
  {
  	next
  }
  if (table(test.mat2$sv_status)[1] < 3 || table(test.mat2$sv_status)[2] < 3)
  {
  	next
  }
  #print(paste0(table(test.mat2$sv_status)[1],"  ", table(test.mat2$sv_status)[2],"  ", nrow(test.mat2)))
  
  ratio <- paste0(sv, "/", no_sv)
  freq <- round((sv/(sv+no_sv))*100, 1)
  freq2 <- round((no_sv/(sv+no_sv))*100, 1)
  
  # Apply frequncy cutoff for sv (set by the user)
  if(opt$fcutoff) {
    if(freq < opt$fcutoff){
      #print("Frequncy less than the user-set threshold!")
      next
    }
  }
  
  # Apply frequency cutoff for no_sv (set to 10%)
  if(freq2 < 10){
    print("no_sv frequncy less than 10% !")
    next
  }
  

  # Run the linear model (glm() default setting is the same as lm() )
  #print(gene_id)
  #print(test.mat2)
  glm <- glm(expression ~ ., data = test.mat2)
  summary <- summary(glm)
  #print(summary)
  
  #extract the pval for sv_status
  estimate <- summary$coefficients[rownames(summary$coefficients) == "sv_statussv", 1]
  error <- summary$coefficients[rownames(summary$coefficients) == "sv_statussv", 2]
  tval <-  summary$coefficients[rownames(summary$coefficients) == "sv_statussv", 3]
  pval <- summary$coefficients[rownames(summary$coefficients) == "sv_statussv", 4]
  
  #print(paste0("Extracted estimate: ", estimate))
  #print(paste0("Extracted error: ", error))
  #print(paste0("Extracted tval: ", tval))
  #print(paste0("Extracted pval: ", pval))
  
  if(as.character(pval) == "NaN") {
    print("p-value is NaN. Too few samples after filtering?")
    next
  }


  if(!(opt$pmt)) {
  # print plotting matrix
  	colnames(test.mat2)[1] <- "fpkm.qn.ns"
  	test.mat2$sample_id <- rownames(test.mat2)
  	plot.mat <- test.mat2
  
  	temp <- fpkm.qn.t[,gene_id]
  	temp <- cbind(names(temp), temp)
  	colnames(temp) <- c("sample_id", "fpkm.qn")
  	temp <- as.data.frame(temp)
  	class(temp[,2]) <- "numeric"
  	plot.mat <- merge(plot.mat, temp, by = "sample_id", all.x = TRUE)

  	temp <- med_devs[,gene_id]
  	temp <- cbind(names(temp), temp)
  	colnames(temp) <- c("sample_id", "fpkm.qn.rle")
  	temp <- as.data.frame(temp)
  	class(temp[,2]) <- "numeric"
  	plot.mat <- merge(plot.mat, temp, by = "sample_id", all.x = TRUE)
  
    write.table(plot.mat, file = paste0(opt$write, "matrix/", gene_id,"_mat.txt"), quote = FALSE, sep = "\t", row.names = FALSE)
  }

  
  Ratio <- c(Ratio, ratio)
  Freq <- c(Freq, freq)
  #print(paste0("SV ratio and percent frequency for gene ", gene_id, ": ", ratio, ", ", freq, "%"))

  Gene <- c(Gene, gene_id)
  Estimate <- c(Estimate, estimate)
  StdError <- c(StdError, error)
  t <- c(t, tval)
  pvalue <- c(pvalue, pval)
  
  
  #print("Running gene_id:")
  #print(tail(Gene,3))
  #print("Running estimate:")
  #print(tail(Estimate,3))
  #print("Running error:")
  #print(tail(StdError,3))
  #print("Running tval:")
  #print(tail(t,3))
  #print("Running pval:")
  #print(tail(pvalue,3))
  
  #print(length(Gene))
  #print(length(Estimate))
  #print(length(StdError))
  #print(length(t))
  #print(length(pvalue))
}
print("Completed normal-score regression.")

if(opt$verbose){print(paste0(i-1, " iterations of the loop have been completed successfully; the loop stalled at i = ", i))}


# Compile results into a table ####
# Make a results table 
results <- cbind(Gene, Ratio, Freq, Estimate, StdError, t, pvalue)
results <- as.data.frame(results)
results$Freq <- as.numeric(results$Freq)
results$Estimate <- as.numeric(results$Estimate)
results$StdError <- as.numeric(results$StdError)
results$t <- as.numeric(results$t)
results$pvalue <- as.numeric(results$pvalue)
#str(results)
#print("Results:")
#head(results)
if(opt$verbose){print(paste0("number of genes tested: ", nrow(results)))}

outputName1 <- paste0(opt$prefix,".sv")
if(opt$pur) {outputName1 <- paste0(outputName1,"_pur")}
if(opt$cn) {outputName1 <- paste0(outputName1,"_cn")}
if(opt$age) {outputName1 <- paste0(outputName1,"_age")}
if(opt$sex) {outputName1 <- paste0(outputName1,"_sex")}
if(opt$type) {outputName1 <- paste0(outputName1,"_type")}
if(opt$PC){outputName1 <- paste0(outputName1,"_PC",opt$npc)}
outputName1 <- paste0(outputName1, ".txt")


if (!(file.exists(opt$dir))) {
  dir.create(opt$dir, recursive = TRUE)
}
if (file.exists(opt$write)) {
	write.table(results, file = paste0(opt$dir,outputName1), quote = FALSE, sep = "\t", row.names = FALSE)
} else {
	dir.create(opt$write, recursive=TRUE)
	write.table(results, file = paste0(opt$dir,outputName1), quote = FALSE, sep = "\t", row.names = FALSE)
}


# Annotate and order genes ####
colnames(results)[1] <- "gene_id"
results <- merge(transcripts, results, by = "gene_id", all.y = TRUE)

# Separate Oncogenes from Suppressors and Perform FDR correction ####
# divide p-value by 2 to calculate one-sided p-value
results$p.onesided <- (results$pvalue)/2

# remove genes with negative Estimate (these are putative tumor suppressors)
#nrow(results)
results.posEstimate <- results[results$Estimate > 0,] #oncogene candidates
#nrow(results.posEstimate)

results.negEstimate <- results[results$Estimate < 0,] #suppressor candidates
#nrow(results.negEstimate)

# Calculate FDR and order
results.posEstimate$fdr.onesided <- p.adjust(results.posEstimate$p.onesided, method = "fdr")
results.negEstimate$fdr.onesided <- p.adjust(results.negEstimate$p.onesided, method = "fdr")

results.posEstimate <- results.posEstimate[order(results.posEstimate$fdr.onesided),]
results.negEstimate <- results.negEstimate[order(results.negEstimate$fdr.onesided),]

if(opt$verbose){
  print("Top oncogene candidates:")
  print(head(results.posEstimate))
  print("Top suppressor candidates:")
  print(head(results.negEstimate))
}

# Output
outputName <- paste0(opt$prefix,".annot_sv")
if(opt$pur) {outputName <- paste0(outputName,"_pur")}
if(opt$cn) {outputName <- paste0(outputName,"_cn")}
if(opt$age) {outputName <- paste0(outputName,"_age")}
if(opt$sex) {outputName <- paste0(outputName,"_sex")}
if(opt$type) {outputName <- paste0(outputName,"_type")}
if(opt$PC){outputName <- paste0(outputName,"_PC",opt$npc)}
outputName.posEstimate <- paste0(outputName, "_posEstimate.txt")
outputName.negEstimate <- paste0(outputName, "_negEstimate.txt")

results.posEstimate <- subset(results.posEstimate, select = -c(TYPE, transcript_id))
results.negEstimate <- subset(results.negEstimate, select = -c(TYPE, transcript_id))

write.table(results.posEstimate, file = paste0(opt$dir,outputName.posEstimate), quote = FALSE, sep = "\t", row.names = FALSE)
write.table(results.negEstimate, file = paste0(opt$dir,outputName.negEstimate), quote = FALSE, sep = "\t", row.names = FALSE)


#save.image(file = "My_Object.RData")
options(warn = oldw)